### PR TITLE
Add support for json output in subkey

### PR DESCRIPTION
**Pull request by hbakkum-dotstar**
Adds support to specify an output type to the subkey generate and inspect subcommands where output type is one of text (i.e. the current output format) or json.

### CODE REVIEW

**Comment by hbakkum-dotstar** [comment]
[clabot:check]

File: bin/utils/subkey/src/main.rs [hunk 1]
```diff
@@ -329,13 +398,24 @@ where
 	if let Some(network) = maybe_network {
 		set_default_ss58_version(network);
 	}
+
+	let maybe_output: Option<OutputType> = match matches.value_of("output").map(|output| {
```
Comment (NikVolf) [inline bin/utils/subkey/src/main.rs]: should not be optional?
If there is none in matches provided, default one is instantiated

File: bin/utils/subkey/src/main.rs [hunk 2]
```diff
@@ -55,50 +74,99 @@ trait Crypto: Sized {
 		uri: &str,
 		password: Option<&str>,
 		network_override: Option<Ss58AddressFormat>,
+		output: Option<OutputType>,
```
Comment (NikVolf) [inline bin/utils/subkey/src/main.rs]: can be just `OutputType` here already

File: bin/utils/subkey/src/main.rs [hunk 3]
```diff
@@ -55,50 +74,97 @@ trait Crypto: Sized {
...
+						Public key (hex): {}\n  \
+						Account ID:       {}\n  \
+						SS58 Address:     {}",
+						uri,
+						format_seed::<Self>(seed),
+						format_public_key::<Self>(public_key.clone()),
+						format_account_id::<Self>(public_key),
+						Self::ss58_from_pair(&pair)
+					);
+				}
```
Comment (NikVolf) [inline bin/utils/subkey/src/main.rs]:
```suggestion
				},
```

File: bin/utils/subkey/src/main.rs [hunk 4]
```diff
@@ -55,50 +74,97 @@ trait Crypto: Sized {
...
+						Public key (hex): {}\n  \
+						Account ID:       {}\n  \
+						SS58 Address:     {}",
+						uri,
+						if let Some(seed) = seed { format_seed::<Self>(seed) } else { "n/a".into() },
+						format_public_key::<Self>(public_key.clone()),
+						format_account_id::<Self>(public_key),
+						Self::ss58_from_pair(&pair)
+					);
+				}
```
Comment (NikVolf) [inline bin/utils/subkey/src/main.rs]:
```suggestion
				},
```

File: bin/utils/subkey/src/main.rs [hunk 5]
```diff
@@ -55,50 +74,97 @@ trait Crypto: Sized {
...
+						Public key (hex):   {}\n  \
+						Account ID:         {}\n  \
+						SS58 Address:       {}",
+						uri,
+						String::from(v),
+						format_public_key::<Self>(public_key.clone()),
+						format_account_id::<Self>(public_key.clone()),
+						public_key.to_ss58check_with_version(v)
+					);
+				}
```
Comment (NikVolf) [inline bin/utils/subkey/src/main.rs]:
```suggestion
				},
```

File: bin/utils/subkey/src/main.rs [hunk 6]
```diff
@@ -55,50 +74,97 @@ trait Crypto: Sized {
...
+					println!("Public Key URI `{}` is account:\n  \
+						Network ID/version: {}\n  \
+						Public key (hex):   {}\n  \
+						Account ID:         {}\n  \
+						SS58 Address:       {}",
+						uri,
+						String::from(v),
+						format_public_key::<Self>(public_key.clone()),
+						format_account_id::<Self>(public_key.clone()),
+						public_key.to_ss58check_with_version(v)
```
Comment (NikVolf) [inline bin/utils/subkey/src/main.rs]:
```suggestion
						public_key.to_ss58check_with_version(v),
```

File: bin/utils/subkey/src/main.rs [hunk 7]
```diff
@@ -55,50 +74,97 @@ trait Crypto: Sized {
...
-			);
+
+			match output {
+				OutputType::Json => {
+					let json = json!({
+						"publicKeyUri": uri,
+						"networkId": String::from(v),
+						"publicKey": format_public_key::<Self>(public_key.clone()),
+						"accountId": format_account_id::<Self>(public_key.clone()),
+						"ss58Address": public_key.to_ss58check_with_version(v)
```
Comment (NikVolf) [inline bin/utils/subkey/src/main.rs]:
```suggestion
						"ss58Address": public_key.to_ss58check_with_version(v),
```

File: bin/utils/subkey/src/main.rs [hunk 8]
```diff
@@ -55,50 +74,97 @@ trait Crypto: Sized {
...
+					println!("Secret Key URI `{}` is account:\n  \
+						Secret seed:      {}\n  \
+						Public key (hex): {}\n  \
+						Account ID:       {}\n  \
+						SS58 Address:     {}",
+						uri,
+						if let Some(seed) = seed { format_seed::<Self>(seed) } else { "n/a".into() },
+						format_public_key::<Self>(public_key.clone()),
+						format_account_id::<Self>(public_key),
+						Self::ss58_from_pair(&pair)
```
Comment (NikVolf) [inline bin/utils/subkey/src/main.rs]:
```suggestion
						Self::ss58_from_pair(&pair),
```

File: bin/utils/subkey/src/main.rs [hunk 9]
```diff
@@ -55,50 +74,97 @@ trait Crypto: Sized {
...
-			);
+
+			match output {
+				OutputType::Json => {
+					let json = json!({
+						"secretKeyUri": uri,
+						"secretSeed": if let Some(seed) = seed { format_seed::<Self>(seed) } else { "n/a".into() },
+						"publicKey": format_public_key::<Self>(public_key.clone()),
+						"accountId": format_account_id::<Self>(public_key),
+						"ss58Address": Self::ss58_from_pair(&pair)
```
Comment (NikVolf) [inline bin/utils/subkey/src/main.rs]:
```suggestion
						"ss58Address": Self::ss58_from_pair(&pair) ,
```

File: bin/utils/subkey/src/main.rs [hunk 10]
```diff
@@ -55,50 +74,97 @@ trait Crypto: Sized {
...
+					println!("Secret phrase `{}` is account:\n  \
+						Secret seed:      {}\n  \
+						Public key (hex): {}\n  \
+						Account ID:       {}\n  \
+						SS58 Address:     {}",
+						uri,
+						format_seed::<Self>(seed),
+						format_public_key::<Self>(public_key.clone()),
+						format_account_id::<Self>(public_key),
+						Self::ss58_from_pair(&pair)
```
Comment (NikVolf) [inline bin/utils/subkey/src/main.rs]:
```suggestion
						Self::ss58_from_pair(&pair),
```

File: bin/utils/subkey/src/main.rs [hunk 11]
```diff
@@ -55,50 +74,97 @@ trait Crypto: Sized {
...
-			);
+			
+			match output {
+				OutputType::Json => {
+					let json = json!({
+						"secretPhrase": uri,
+						"secretSeed": format_seed::<Self>(seed),
+						"publicKey": format_public_key::<Self>(public_key.clone()),
+						"accountId": format_account_id::<Self>(public_key),
+						"ss58Address": Self::ss58_from_pair(&pair)
```
Comment (NikVolf) [inline bin/utils/subkey/src/main.rs]:
```suggestion
						"ss58Address": Self::ss58_from_pair(&pair),
```

File: bin/utils/subkey/src/main.rs [hunk 12]
```diff
@@ -37,6 +38,24 @@ use std::{
...
+}
+
+impl<'a> TryFrom<&'a str> for OutputType {
+	type Error = ();
+
+	fn try_from(s: &'a str) -> Result<OutputType, ()> {
+		match s {
+			"json" => Ok(OutputType::Json),
+			"text" => Ok(OutputType::Text),
+			_      => Err(())
```
Comment (NikVolf) [inline bin/utils/subkey/src/main.rs]:
```suggestion
			_      => Err(()),
```

File: bin/utils/subkey/src/main.rs [hunk 13]
```diff
@@ -37,6 +38,24 @@ use std::{
...
 
+enum OutputType {
+	Json,
+	Text
+}
+
+impl<'a> TryFrom<&'a str> for OutputType {
+	type Error = ();
+
+	fn try_from(s: &'a str) -> Result<OutputType, ()> {
```
Comment (NikVolf) [inline bin/utils/subkey/src/main.rs]:
```suggestion
	fn try_from(s: &'a str) -> Result<OutputType, &'static str> {
```

Comment (NikVolf) [reply bin/utils/subkey/src/main.rs]: But this can probably left as it is, since `()` is mapped below anyway

Comment (hbakkum-dotstar) [reply bin/utils/subkey/src/main.rs]: Cool - will leave this as is then.

File: bin/utils/subkey/src/main.rs [hunk 14]
```diff
@@ -37,6 +38,24 @@ use std::{
...
+		match s {
+			"json" => Ok(OutputType::Json),
+			"text" => Ok(OutputType::Text),
+			_      => Err(()),
```
Comment (bkchr) [inline bin/utils/subkey/src/main.rs]:
```suggestion
			_ => Err(()),
```

File: bin/utils/subkey/src/main.rs [hunk 15]
```diff
@@ -329,13 +396,24 @@ where
 	if let Some(network) = maybe_network {
 		set_default_ss58_version(network);
 	}
+
+	let output: OutputType = match matches.value_of("output").map(|output| {
+		output
+			.try_into()
+			.map_err(|_| Error::Static("Invalid output name. See --help for available outputs."))
```
Comment (bkchr) [inline bin/utils/subkey/src/main.rs]: You should move this error down to the line `405`.

Than you can change the map to `map(TryInto::tryinto)`

File: bin/utils/subkey/src/main.rs [hunk 16]
```diff
@@ -55,50 +74,97 @@ trait Crypto: Sized {
...
+			match output {
+				OutputType::Json => {
+					let json = json!({
+						"secretPhrase": uri,
+						"secretSeed": format_seed::<Self>(seed),
+						"publicKey": format_public_key::<Self>(public_key.clone()),
+						"accountId": format_account_id::<Self>(public_key),
+						"ss58Address": Self::ss58_from_pair(&pair),
+					});
+					println!("{}", serde_json::to_string_pretty(&json).unwrap());
```
Comment (bkchr) [inline bin/utils/subkey/src/main.rs]:
```suggestion
					println!("{}", serde_json::to_string_pretty(&json).expect("Json pretty print failed"));
```

File: bin/utils/subkey/src/main.rs [hunk 17]
```diff
@@ -55,50 +74,97 @@ trait Crypto: Sized {
...
+			match output {
+				OutputType::Json => {
+					let json = json!({
+						"secretKeyUri": uri,
+						"secretSeed": if let Some(seed) = seed { format_seed::<Self>(seed) } else { "n/a".into() },
+						"publicKey": format_public_key::<Self>(public_key.clone()),
+						"accountId": format_account_id::<Self>(public_key),
+						"ss58Address": Self::ss58_from_pair(&pair),
+					});
+					println!("{}", serde_json::to_string_pretty(&json).unwrap());
```
Comment (bkchr) [inline bin/utils/subkey/src/main.rs]:
```suggestion
					println!("{}", serde_json::to_string_pretty(&json).expect("Json pretty print failed"));
```

File: bin/utils/subkey/src/main.rs [hunk 18]
```diff
@@ -55,50 +74,97 @@ trait Crypto: Sized {
...
+			match output {
+				OutputType::Json => {
+					let json = json!({
+						"publicKeyUri": uri,
+						"networkId": String::from(v),
+						"publicKey": format_public_key::<Self>(public_key.clone()),
+						"accountId": format_account_id::<Self>(public_key.clone()),
+						"ss58Address": public_key.to_ss58check_with_version(v),
+					});
+					println!("{}", serde_json::to_string_pretty(&json).unwrap());
```
Comment (bkchr) [inline bin/utils/subkey/src/main.rs]:
```suggestion
					println!("{}", serde_json::to_string_pretty(&json).expect("Json pretty print failed"));
```

File: bin/utils/subkey/src/main.rs [hunk 19]
```diff
@@ -37,6 +38,24 @@ use std::{
 mod rpc;
 mod vanity;
 
+enum OutputType {
+	Json,
+	Text
```
Comment (marcio-diaz) [inline bin/utils/subkey/src/main.rs]:
```suggestion
	Text,
```

File: bin/utils/subkey/src/main.rs [hunk 20]
```diff
@@ -55,50 +74,97 @@ trait Crypto: Sized {
...
-				public_key.to_ss58check_with_version(v)
-			);
+
+			match output {
+				OutputType::Json => {
+					let json = json!({
+						"publicKeyUri": uri,
+						"networkId": String::from(v),
+						"publicKey": format_public_key::<Self>(public_key.clone()),
+						"accountId": format_account_id::<Self>(public_key.clone()),
```
Comment (marcio-diaz) [inline bin/utils/subkey/src/main.rs]:
```suggestion
						"accountId": format_account_id::<Self>(public_key),
```

File: bin/utils/subkey/src/main.rs [hunk 21]
```diff
@@ -55,50 +74,97 @@ trait Crypto: Sized {
...
+				OutputType::Text => {
+					println!("Public Key URI `{}` is account:\n  \
+						Network ID/version: {}\n  \
+						Public key (hex):   {}\n  \
+						Account ID:         {}\n  \
+						SS58 Address:       {}",
+						uri,
+						String::from(v),
+						format_public_key::<Self>(public_key.clone()),
+						format_account_id::<Self>(public_key.clone()),
```
Comment (marcio-diaz) [inline bin/utils/subkey/src/main.rs]:
```suggestion
						format_account_id::<Self>(public_key),
```

File: bin/utils/subkey/secret [hunk 22]
```diff
@@ -0,0 +1 @@
+�.o��������R�c�3J}��8���b�}&Ai
```
Comment (marcio-diaz) [inline bin/utils/subkey/secret]: remove it?

Comment (hbakkum-dotstar) [reply bin/utils/subkey/secret]: heh, yes accidentally committed while testing on another branch - deleted

**Pull request merged**